src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -146,13 +161,20 @@ public function onKernelResponse(FilterResponseEvent $event)
         if ((null === $token = $this->tokenStorage->getToken()) || $this->trustResolver->isAnonymous($token)) {
             if ($request->hasPreviousSession()) {
                 $session->remove($this->sessionKey);
+                $session->remove($this->refreshableRolesSessionKey);
             }
         } else {
             $session->set($this->sessionKey, serialize($token));
 
             if (null !== $this->logger) {
                 $this->logger->debug('Stored the security token in the session.', array('key' => $this->sessionKey));
             }
+
+            // if the token changed, then re-set the refreshable key
+            if ($token !== $this->refreshedToken) {
+                $shouldUpdateRoles = $token instanceof RefreshableRolesTokenInterface && $token->shouldUpdateRoles();
+                $session->set($this->refreshableRolesSessionKey, $shouldUpdateRoles);
+            }
```

Comment:
This was a tricky part. The return value from `shouldUpdateRoles()` is not serialized to the session (i.e. it's not in `AbstractToken::serialize()`. This is done for BC: we can't change `serialize()` & `unserialize()` to include this.

So, we instead store this value in a different session key. This logic only sets it if the token has *changed*. Basically, we want to set this session key during the original request when authentication occurred (that's when we know that `shouldUpdateRoles()` is accurate). By checking to see if the token changed, we're effectively accomplishing that.

This was the weirdest / hackiest part of this PR. I'm open to suggestions.

Reply:
If your upgrading there is a big chance you already need to clear existing session data. So this change should be acceptable.